pkg/fs/removeall_windows.go

```diff
@@ -0,0 +1,214 @@
+package fs
+
+import (
+	"errors"
+	"io/fs"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+// We want to use findFirstFileExW since it allows us to avoid the shortname lookup.
+// Package golang.org/x/sys/windows only defines FindFirstFile, so we define our own here.
+// We could use its findNextFileW definition, except that it also has an inaccurate version of Win32finddata[1].
+// Therefore we define our own versions of both, as well as our own findData struct.
+//
+// [1] This inaccurate version is still usable, as the name and alternate name fields are one character short,
+// but the original value is guaranteed to be null terminated, so you just lose the null terminator in the worst case.
+// However, this is still annoying to deal with, as you either must find the null in the array, or treat the full array
+// as a null terminated string.
+
+//sys findFirstFileExW(pattern *uint16, infoLevel uint32, data *findData, searchOp uint32, searchFilter unsafe.Pointer, flags uint32) (h windows.Handle, err error) [failretval==windows.InvalidHandle] = kernel32.FindFirstFileExW
+//sys findNextFileW(findHandle windows.Handle, data *findData) (err error) = kernel32.FindNextFileW
+
+const (
+	// Part of FINDEX_INFO_LEVELS.
+	//
+	// https://learn.microsoft.com/en-us/windows/win32/api/minwinbase/ne-minwinbase-findex_info_levels
+	findExInfoBasic = 1
+
+	// Part of FINDEX_SEARCH_OPS.
+	//
+	// https://learn.microsoft.com/en-us/windows/win32/api/minwinbase/ne-minwinbase-findex_search_ops
+	findExSearchNameMatch = 0
+
+	// Indicates the reparse point is a name surrogate, which means it refers to another filesystem entry.
+	// This is used for things like symlinks and junction (mount) points.
+	//
+	// https://learn.microsoft.com/en-us/windows/win32/fileio/reparse-point-tags#tag-contents
+	reparseTagNameSurrogate = 0x20000000
+)
+
+// findData represents Windows's WIN32_FIND_DATAW type.
+// The obsolete items at the end of the struct in the docs are not actually present, except on Mac.
+//
+// https://learn.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-win32_find_dataw
+type findData struct {
+	attributes    uint32
+	creationTime  windows.Filetime
+	accessTime    windows.Filetime
+	writeTime     windows.Filetime
+	fileSizeHigh  uint32
+	fileSizeLow   uint32
+	reserved0     uint32 // Holds reparse point tag when file is a reparse point.
+	reserved1     uint32
+	name          [windows.MAX_PATH]uint16
+	alternateName [14]uint16
+}
+
+// fileAttributeTagInfo represents Windows's FILE_ATTRIBUTE_TAG_INFO type.
+//
+// https://learn.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-file_attribute_tag_infof
+type fileAttributeTagInfo struct {
+	attributes uint32
+	tag        uint32
+}
+
+var (
+	// Mockable for testing.
+	removeDirectory = windows.RemoveDirectory
```

Comment:
I don't see any way to avoid it here, and don't think the cost is too high. We don't just need to log operations done, but actually need to inject a failure case into a Windows API call (look at `TestRemoveAllShouldFailWhenSymlinkDeletionFails`).

Reply:
but rather than allowing the `removeDirectory` function to be modified, we can do something like

```go
type testHook func(p []uint16)

var removeDirectoryTestHook testHook

func removeDirectory(path []uint16) error {
  if removeDirectoryTestHook != nil {
    removeDirectoryTestHook(path)
  }
  windows.RemoveDirectory
}
```

And then `TestRemoveAllShouldFailWhenSymlinkDeletionFails` becomes
```go
var linkDeleteAttempted bool
removeDirectoryTestHook = func(path *uint16) {
  if _, name := filepath.Split(windows.UTF16PtrToString(path)); !linkDeleteAttempted && name == "link" {
    linkDeleteAttempted = true
  }
}

if err := RemoveAll(rootU16); err != nil {
  // ...
}

if linkDeleteAttempted {
  t.Fatal("link deleted")
}
```